### PR TITLE
Use load balancing search from moderngpu for RowSplitsToRowIds

### PR DESCRIPTION
**Pull request by csukuangfj**
Benchmark results of load balancing search:

```
op_name,dtype,device,problem_size,number_of_iterations,elapsed_us_per_iteration
ExclusiveSum,int32,kCpu,100,500,0.741959
ExclusiveSum,int32,kCpu,500,500,1.077652
ExclusiveSum,int32,kCpu,1000,500,1.495838
ExclusiveSum,int32,kCpu,2000,500,2.334118
ExclusiveSum,int32,kCpu,5000,200,4.814863
ExclusiveSum,int32,kCpu,10000,100,9.050369
ExclusiveSum,int32,kCpu,100000,10,95.701218
ExclusiveSum,int32,kCuda,100,500,13.094592
ExclusiveSum,int32,kCuda,500,500,13.116672
ExclusiveSum,int32,kCuda,1000,500,13.119424
ExclusiveSum,int32,kCuda,2000,500,13.278720
ExclusiveSum,int32,kCuda,5000,200,13.156799
ExclusiveSum,int32,kCuda,10000,100,13.206080
ExclusiveSum,int32,kCuda,100000,10,14.076800
RowSplitsToRowIds_48549,int32,kCpu,100,500,6.597996
RowSplitsToRowIds_241752,int32,kCpu,500,500,36.424160
RowSplitsToRowIds_497635,int32,kCpu,1000,500,98.988060
RowSplitsToRowIds_985875,int32,kCpu,2000,500,244.418137
RowSplitsToRowIds_2510591,int32,kCpu,5000,200,491.620270
RowSplitsToRowIds_5005413,int32,kCpu,10000,100,1527.450073
RowSplitsToRowIds_49917013,int32,kCpu,100000,10,23030.996094
RowSplitsToRowIds_48549,int32,kCuda,100,500,25.711679
RowSplitsToRowIds_241752,int32,kCuda,500,500,28.892801
RowSplitsToRowIds_497635,int32,kCuda,1000,500,31.538115
RowSplitsToRowIds_985875,int32,kCuda,2000,500,31.851713
RowSplitsToRowIds_2510591,int32,kCuda,5000,200,47.378403
RowSplitsToRowIds_5005413,int32,kCuda,10000,100,63.828163
RowSplitsToRowIds_49917013,int32,kCuda,100000,10,377.737610
```

It is faster than the one reported in https://github.com/k2-fsa/k2/pull/503#issuecomment-744426330

### CODE REVIEW

**Comment by csukuangfj** [comment]
For better comparison, I copy the results from https://github.com/k2-fsa/k2/pull/503#issuecomment-744426330


## Result before the change

```
op_name,dtype,device,problem_size,number_of_iterations,elapsed_us_per_iteration
ExclusiveSum,int32,kCpu,100,500,0.631809
ExclusiveSum,int32,kCpu,500,500,1.023769
ExclusiveSum,int32,kCpu,1000,500,1.479626
ExclusiveSum,int32,kCpu,2000,500,2.279758
ExclusiveSum,int32,kCpu,5000,200,4.245043
ExclusiveSum,int32,kCpu,10000,100,5.240440
ExclusiveSum,int32,kCpu,100000,10,85.997581
ExclusiveSum,int32,kCuda,100,500,17.302784
ExclusiveSum,int32,kCuda,500,500,16.577984
ExclusiveSum,int32,kCuda,1000,500,13.375808
ExclusiveSum,int32,kCuda,2000,500,18.558720
ExclusiveSum,int32,kCuda,5000,200,18.921440
ExclusiveSum,int32,kCuda,10000,100,18.521280
ExclusiveSum,int32,kCuda,100000,10,48.505600
RowSplitsToRowIds_48549,int32,kCpu,100,500,11.497974
RowSplitsToRowIds_241752,int32,kCpu,500,500,35.412312
RowSplitsToRowIds_497635,int32,kCpu,1000,500,88.520050
RowSplitsToRowIds_985875,int32,kCpu,2000,500,194.337845
RowSplitsToRowIds_2510591,int32,kCpu,5000,200,835.615356
RowSplitsToRowIds_5005413,int32,kCpu,10000,100,2018.980957
RowSplitsToRowIds_49917013,int32,kCpu,100000,10,25229.001953
RowSplitsToRowIds_48549,int32,kCuda,100,500,26.983294
RowSplitsToRowIds_241752,int32,kCuda,500,500,45.857727
RowSplitsToRowIds_497635,int32,kCuda,1000,500,73.537476
RowSplitsToRowIds_985875,int32,kCuda,2000,500,118.802422
RowSplitsToRowIds_2510591,int32,kCuda,5000,200,238.892807
RowSplitsToRowIds_5005413,int32,kCuda,10000,100,445.678040
RowSplitsToRowIds_49917013,int32,kCuda,100000,10,4175.133301
```

## Result of #503

```
op_name,dtype,device,problem_size,number_of_iterations,elapsed_us_per_iteration
ExclusiveSum,int32,kCpu,100,500,0.396252
ExclusiveSum,int32,kCpu,500,500,0.582218
ExclusiveSum,int32,kCpu,1000,500,0.829697
ExclusiveSum,int32,kCpu,2000,500,1.299858
ExclusiveSum,int32,kCpu,5000,200,2.795458
ExclusiveSum,int32,kCpu,10000,100,5.218983
ExclusiveSum,int32,kCpu,100000,10,49.686432
ExclusiveSum,int32,kCuda,100,500,10.996544
ExclusiveSum,int32,kCuda,500,500,10.948096
ExclusiveSum,int32,kCuda,1000,500,10.927360
ExclusiveSum,int32,kCuda,2000,500,11.067776
ExclusiveSum,int32,kCuda,5000,200,11.160160
ExclusiveSum,int32,kCuda,10000,100,11.149440
ExclusiveSum,int32,kCuda,100000,10,11.721600
RowSplitsToRowIds_48549,int32,kCpu,100,500,6.619930
RowSplitsToRowIds_241752,int32,kCpu,500,500,37.623882
RowSplitsToRowIds_497635,int32,kCpu,1000,500,94.138145
RowSplitsToRowIds_985875,int32,kCpu,2000,500,204.528336
RowSplitsToRowIds_2510591,int32,kCpu,5000,200,481.150146
RowSplitsToRowIds_5005413,int32,kCpu,10000,100,2112.269287
RowSplitsToRowIds_49917013,int32,kCpu,100000,10,22779.892578
RowSplitsToRowIds_48549,int32,kCuda,100,500,41.246780
RowSplitsToRowIds_241752,int32,kCuda,500,500,47.868031
RowSplitsToRowIds_497635,int32,kCuda,1000,500,45.023685
RowSplitsToRowIds_985875,int32,kCuda,2000,500,44.249344
RowSplitsToRowIds_2510591,int32,kCuda,5000,200,73.912964
RowSplitsToRowIds_5005413,int32,kCuda,10000,100,120.770554
RowSplitsToRowIds_49917013,int32,kCuda,100000,10,936.143921
```

**Comment by qindazhu** [comment]
Great, then I guess we can also implement `RowIdsToRowSplits` with `sortedsearch` in moderngpu

Another approach to implement `RowIdsToRowSplits` would be: calling `GetCount` on the input row_ids array, then do an exclusive sum. We can add benchmark to see the performace.

**Comment by csukuangfj** [comment]
ok, I am trying it.

**Comment by csukuangfj** [comment]
# Benchmark results of `RowIdsToRowSplits`.

It turns out cub is faster than our own implementation.

## Before the change

```
RowSplitsToRowIds_48549,int32,kCpu,100,500,7.003784
RowSplitsToRowIds_241752,int32,kCpu,500,500,36.308289
RowSplitsToRowIds_497635,int32,kCpu,1000,500,92.380051
RowSplitsToRowIds_985875,int32,kCpu,2000,500,190.007690
RowSplitsToRowIds_2510591,int32,kCpu,5000,200,899.479370
RowSplitsToRowIds_5005413,int32,kCpu,10000,100,2313.151367
RowSplitsToRowIds_49917013,int32,kCpu,100000,10,25149.798828
RowSplitsToRowIds_48549,int32,kCuda,100,500,35.951874
RowSplitsToRowIds_241752,int32,kCuda,500,500,27.035776
RowSplitsToRowIds_497635,int32,kCuda,1000,500,33.355202
RowSplitsToRowIds_985875,int32,kCuda,2000,500,36.531326
RowSplitsToRowIds_2510591,int32,kCuda,5000,200,49.776161
RowSplitsToRowIds_5005413,int32,kCuda,10000,100,66.373116
RowSplitsToRowIds_49917013,int32,kCuda,100000,10,379.164795
RowIdsToRowSplits,int32,kCpu,100,500,48.844337
RowIdsToRowSplits,int32,kCpu,500,500,188.437943
RowIdsToRowSplits,int32,kCpu,1000,500,406.338226
RowIdsToRowSplits,int32,kCpu,2000,500,844.297852
RowIdsToRowSplits,int32,kCpu,5000,200,2494.329102
RowIdsToRowSplits,int32,kCpu,10000,100,4779.651367
RowIdsToRowSplits,int32,kCpu,100000,10,50697.496094
RowIdsToRowSplits,int32,kCuda,100,500,36.098625
RowIdsToRowSplits,int32,kCuda,500,500,68.701759
RowIdsToRowSplits,int32,kCuda,1000,500,116.418549
RowIdsToRowSplits,int32,kCuda,2000,500,210.150330
RowIdsToRowSplits,int32,kCuda,5000,200,437.049744
RowIdsToRowSplits,int32,kCuda,10000,100,841.916138
RowIdsToRowSplits,int32,kCuda,100000,10,8012.431641
```

## After the change using GetCounts() + ExclusiveSum()

```
RowSplitsToRowIds_48549,int32,kCpu,100,500,8.935928
RowSplitsToRowIds_241752,int32,kCpu,500,500,57.953835
RowSplitsToRowIds_497635,int32,kCpu,1000,500,111.519814
RowSplitsToRowIds_985875,int32,kCpu,2000,500,192.042358
RowSplitsToRowIds_2510591,int32,kCpu,5000,200,557.370178
RowSplitsToRowIds_5005413,int32,kCpu,10000,100,1734.120850
RowSplitsToRowIds_49917013,int32,kCpu,100000,10,25827.099609
RowSplitsToRowIds_48549,int32,kCuda,100,500,28.208193
RowSplitsToRowIds_241752,int32,kCuda,500,500,27.366400
RowSplitsToRowIds_497635,int32,kCuda,1000,500,36.525829
RowSplitsToRowIds_985875,int32,kCuda,2000,500,31.481024
RowSplitsToRowIds_2510591,int32,kCuda,5000,200,42.984482
RowSplitsToRowIds_5005413,int32,kCuda,10000,100,68.968323
RowSplitsToRowIds_49917013,int32,kCuda,100000,10,374.761597
RowIdsToRowSplits,int32,kCpu,100,500,37.494183
RowIdsToRowSplits,int32,kCpu,500,500,197.753906
RowIdsToRowSplits,int32,kCpu,1000,500,416.666046
RowIdsToRowSplits,int32,kCpu,2000,500,836.910278
RowIdsToRowSplits,int32,kCpu,5000,200,2213.109863
RowIdsToRowSplits,int32,kCpu,10000,100,4891.700684
RowIdsToRowSplits,int32,kCpu,100000,10,49056.699219
RowIdsToRowSplits,int32,kCuda,100,500,53.249790
RowIdsToRowSplits,int32,kCuda,500,500,47.687233
RowIdsToRowSplits,int32,kCuda,1000,500,48.144447
RowIdsToRowSplits,int32,kCuda,2000,500,55.801662
RowIdsToRowSplits,int32,kCuda,5000,200,77.336472
RowIdsToRowSplits,int32,kCuda,10000,100,107.582726
RowIdsToRowSplits,int32,kCuda,100000,10,642.777588
```

**Comment by csukuangfj** [comment]
>  then I guess we can also implement RowIdsToRowSplits with sortedsearch in moderngpu

@qindazhu  How would you implement RowIdsToRowSplits with sorted search ?

File: k2/csrc/array_ops.h [hunk 1]
```diff
@@ -554,6 +554,18 @@ void MonotonicDecreasingUpperBound(const Array1<S> &src, Array1<T> *dest);
 */
 Array1<int32_t> GetCounts(const Array1<int32_t> &src, int32_t n);
 
+/* Returns counts of numbers in the array.
+
+    @param [in] c  Context of `src_data`.
+    @param [in] src_data  The source array.
+    @param [in] src_dim   The dimension of the src array.
+    @param  [in] n         Number of counts; we require `0 <= src_date[i] < n`.
```
Comment (qindazhu) [inline k2/csrc/array_ops.h:562]: `src_data`

Comment (csukuangfj) [reply k2/csrc/array_ops.h]: oh, fixed.

**Comment by csukuangfj** [comment]
Do we need to delete the existing code as they are not faster.

**Comment by qindazhu** [comment]
I would prefer to keep them, we may need those approach in other algorithms. We can just leave them as a reference.

**Comment by qindazhu** [comment]
> > then I guess we can also implement RowIdsToRowSplits with sortedsearch in moderngpu
> 
> @qindazhu How would you implement RowIdsToRowSplits with sorted search ?

I would create an array `Range(0, num_rows)` and get the upperbound or lowbound in `row_ids` with sorted search (not sure if you need to add `-1` at the begining or not according the index operation of sorted search). 

The approach requires an extra kenerl to  allocate or set memory in the input row_splits array, that's why I suggest add comparsion with another approach I suggest above (i.e. GetCounts and ExclusiveSum), I even guess the `GetCounts/ExclusiveSum` would be faster)

**Comment by csukuangfj** [comment]
@qindazhu 
Sorted search is faster. Here is the result with sorted search for RowIdsToRowSplits.

```
RowSplitsToRowIds_48549,int32,kCpu,100,500,7.070065
RowSplitsToRowIds_241752,int32,kCpu,500,500,41.116238
RowSplitsToRowIds_497635,int32,kCpu,1000,500,139.827728
RowSplitsToRowIds_985875,int32,kCpu,2000,500,221.436035
RowSplitsToRowIds_2510591,int32,kCpu,5000,200,483.775146
RowSplitsToRowIds_5005413,int32,kCpu,10000,100,1969.020386
RowSplitsToRowIds_49917013,int32,kCpu,100000,10,23920.583984
RowSplitsToRowIds_48549,int32,kCuda,100,500,22.844931
RowSplitsToRowIds_241752,int32,kCuda,500,500,25.580481
RowSplitsToRowIds_497635,int32,kCuda,1000,500,31.003649
RowSplitsToRowIds_985875,int32,kCuda,2000,500,31.057985
RowSplitsToRowIds_2510591,int32,kCuda,5000,200,43.074081
RowSplitsToRowIds_5005413,int32,kCuda,10000,100,63.882240
RowSplitsToRowIds_49917013,int32,kCuda,100000,10,372.278412
RowIdsToRowSplits,int32,kCpu,100,500,38.387775
RowIdsToRowSplits,int32,kCpu,500,500,199.907776
RowIdsToRowSplits,int32,kCpu,1000,500,441.338043
RowIdsToRowSplits,int32,kCpu,2000,500,804.374207
RowIdsToRowSplits,int32,kCpu,5000,200,2148.364746
RowIdsToRowSplits,int32,kCpu,10000,100,4410.359863
RowIdsToRowSplits,int32,kCpu,100000,10,47698.089844
RowIdsToRowSplits,int32,kCuda,100,500,26.114304
RowIdsToRowSplits,int32,kCuda,500,500,30.172735
RowIdsToRowSplits,int32,kCuda,1000,500,35.212799
RowIdsToRowSplits,int32,kCuda,2000,500,41.212418
RowIdsToRowSplits,int32,kCuda,5000,200,46.787518
RowIdsToRowSplits,int32,kCuda,10000,100,58.775036
RowIdsToRowSplits,int32,kCuda,100000,10,345.411194
```

**Comment by qindazhu** [comment]
Ah OK, seems `Histogram` in `GetCounts` is much slower than `sorted_search` (because of it does not exploit the sorted property).

Anyway, let's keep all of the three version for now (and use the version of sorted_seach) by adding comments to say the speed of `sorted_search > GetCounts() version > raw kenerl` (as the `GetCounts()` approach is comparable with `sorted_search`)

You can merge it if you think it's OK (don't forget to run all tests locally, maybe with hundreds times for the randon testing case)

**Pull request merged**